### ISSUE TITLE
Calling parent in onBeforeWrite and onAfterWrite

### ISSUE THREAD

**Issue by jyrkij** [silverstripe/silverstripe-subsites#327]
Shouldn't `parent::onBeforeWrite();` and `parent::onAfterWrite();` be present on all overrides of the respective methods? Not having them there disables extensions on those methods. I'm not sure if it has any other effects.

**Comment by robbieaverill** [comment]
Hi @jyrkij - which part of the code are you referring to?

**Comment by jyrkij** [comment]
Sorry for not clearing that out! Here's a list of methods missing parent calls:

- https://github.com/silverstripe/silverstripe-subsites/blob/784c3fca5e622b675da34bb94625874688fd50b4/src/Model/SubsiteDomain.php#L97-L100
- https://github.com/silverstripe/silverstripe-subsites/blob/784c3fca5e622b675da34bb94625874688fd50b4/src/Extension/GroupSubsites.php#L198-L205
- https://github.com/silverstripe/silverstripe-subsites/blob/784c3fca5e622b675da34bb94625874688fd50b4/src/Extension/GroupSubsites.php#L207-L215

And this one has the parent call in place:

- https://github.com/silverstripe/silverstripe-subsites/blob/0ebf9509bf1f2db29c81e93c5043ec9a30e8aeda/src/Model/Subsite.php#L809-L813

**Comment by robbieaverill** [comment]
Ok, thank you.

Extensions hooks are all called individually, and the DataExtension parent methods don't define any logic, so calling `parent::onAfterWrite` or `parent::onBeforeWrite` for example isn't necessary.

Are you experiencing issues with this, or was it simply a question?

**Comment by jyrkij** [comment]
Yes, I'm experiencing issues. I have an extension for `Subsite` and `SubsiteDomain` and need to hook to `onAfterWrite` on both. As `SubsiteDomain` doesn't call `parent::onAfterWrite();` the hooks are never called.

I'm using latest (SS 4) references here, but we are still stuck on SS 3. The principles and issues are the same though.

`DataObject` (which both classes extend from) says in comments that parent methods should be called:

```php
    /**
     * Event handler called before writing to the database.
     * You can overload this to clean up or otherwise process data before writing it to the
     * database.  Don't forget to call parent::onBeforeWrite(), though!
     *
     * This called after {@link $this->validate()}, so you can be sure that your data is valid.
     *
     * @uses DataExtension->onBeforeWrite()
     */
    protected function onBeforeWrite()
    {
        $this->brokenOnWrite = false;
        $dummy = null;
        $this->extend('onBeforeWrite', $dummy);
    }
    /**
     * Event handler called after writing to the database.
     * You can overload this to act upon changes made to the data after it is written.
     * $this->changed will have a record
     * database.  Don't forget to call parent::onAfterWrite(), though!
     *
     * @uses DataExtension->onAfterWrite()
     */
    protected function onAfterWrite()
    {
        $dummy = null;
        $this->extend('onAfterWrite', $dummy);
    }
```
https://github.com/silverstripe/silverstripe-framework/blob/4.0/src/ORM/DataObject.php#L1025-L1054

`onBeforeWrite` has even a check: `$brokenOnWrite`

```php
    /**
     * Prepare an object prior to write
     *
     * @throws ValidationException
     */
    protected function preWrite()
    {
        // Validate this object
        if ($writeException = $this->validateWrite()) {
            // Used by DODs to clean up after themselves, eg, Versioned
            $this->invokeWithExtensions('onAfterSkippedWrite');
            throw $writeException;
        }
        // Check onBeforeWrite
        $this->brokenOnWrite = true;
        $this->onBeforeWrite();
        if ($this->brokenOnWrite) {
            user_error(static::class . " has a broken onBeforeWrite() function."
                . " Make sure that you call parent::onBeforeWrite().", E_USER_ERROR);
        }
    }
```

https://github.com/silverstripe/silverstripe-framework/blob/4.0/src/ORM/DataObject.php#L1169-L1190

And after re-checking, `GroupSubsite` is an extension itself so there calling parent shouldn't be necessary. So the missing call is only on `SubsiteDomain`.

**Comment by robbieaverill** [comment]
Ah, OK - DataObjects should be calling their parent methods. Would you be keen to make a pull request to add them to the SS3 version? We can discuss further with the core team once changes are up for review, just to make sure this doesn't introduce any _more_ bugs :D

**Comment by jyrkij** [comment]
I'll create a pull request but unfortunately I don't have the possibility to run tests. They're run automatically though?

**Comment by robbieaverill** [comment]
Yep, if anything (covered by existing tests) breaks from your change then we'll know straight away

**Comment by dhensby** [comment]
IMO, even if the parent stubs don't do anything now it doesn't mean they never will and we should be calling them ideally.

**Comment by jyrkij** [comment]
Agree on calling all parents. Still, they aren't absolutely necessary for now...

**Comment by robbieaverill** [comment]
Fixed in #328 - will be in the next 1.3.x release. Thanks for contributing @jyrkij !
